test/extended/router/gatewayapicontroller.go

```diff
@@ -370,6 +360,62 @@ func buildGateway(name, namespace, gcname, fromNs, domain string) *gatewayapiv1.
 	}
 }
 
+func waitForLoadbalancerReady(oc *exutil.CLI, gw, gwServiceName string) {
+	// check gateway LB service, note that External-IP might be hostname (AWS) or IP (Azure/GCP)
+	var lbAddress string
+	gwapiClient := gatewayapiclientset.NewForConfigOrDie(oc.AdminConfig())
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 10*time.Minute, false, func(context context.Context) (bool, error) {
+		lbService, err := oc.AdminKubeClient().CoreV1().Services("openshift-ingress").Get(context, gwServiceName, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		if lbService.Status.LoadBalancer.Ingress[0].Hostname != "" {
+			lbAddress = lbService.Status.LoadBalancer.Ingress[0].Hostname
+		} else {
+			lbAddress = lbService.Status.LoadBalancer.Ingress[0].IP
+		}
+		e2e.Logf("The load balancer External-IP is: %v", lbAddress)
+
+		gwlist, haerr := gwapiClient.GatewayV1().Gateways("openshift-ingress").Get(context, gw, metav1.GetOptions{})
+		e2e.Logf("The gateway hostname address is %v ", gwlist.Status.Addresses[0].Value)
+		o.Expect(haerr).NotTo(o.HaveOccurred())
+		o.Expect(lbAddress).To(o.Equal(gwlist.Status.Addresses[0].Value))
+		return true, nil
+	})
+	o.Expect(err).NotTo(o.HaveOccurred(), "The LB service is not either up nor same with the gateway hostname address")
+}
+
+// assertDNSRecordStatus polls until the DNSRecord's status in the default operand namespace is True.
+func assertDNSRecordStatus(oc *exutil.CLI, gatewayName string) {
+	// find the DNS Record and confirm its zone status is True
+	gatewayDNSRecord := &operatoringressv1.DNSRecord{}
+	err := wait.PollUntilContextTimeout(context.Background(), 2*time.Second, 10*time.Minute, false, func(context context.Context) (bool, error) {
+		gatewayDNSRecords, err := oc.AdminIngressClient().IngressV1().DNSRecords("openshift-ingress").List(context, metav1.ListOptions{})
+		if err != nil {
+			e2e.Logf("Failed to list DNS records for gateway %q, retrying...", gatewayName)
+			return false, nil
+		}
+
+		// get the desired DNS records of the given gateway
+		for count, items := range gatewayDNSRecords.Items {
+			if gatewayDNSRecords.Items[count].Labels["gateway.networking.k8s.io/gateway-name"] == gatewayName {
+				gatewayDNSRecord = &items
+				break
+			}
+		}
+
+		// checking the gateway DNS record status
+		for _, zone := range gatewayDNSRecord.Status.Zones {
+			for _, condition := range zone.Conditions {
+				if condition.Type == "Published" && condition.Status == "True" {
+					return true, nil
+				}
+			}
+		}
+		e2e.Logf("DNS record %q is not ready, retrying...", gatewayDNSRecord.Name)
+		return false, nil
+	})
+	o.Expect(err).NotTo(o.HaveOccurred(), "Timed out waiting for DNSRecord %q to become ready", gatewayDNSRecord.Name)
```

Comment:
We may fail to get the expected DNSRecord in which case `gatewayDNSRecord` will remain empty. This will make the error message look odd (with empty string).
```suggestion
	o.Expect(err).NotTo(o.HaveOccurred(), "Timed out waiting for gateway %q DNSRecord to become ready", gatewayName)
```

Reply:
done